scheduler/core/programprovider/gpp/gppprogramprovider.py

```diff
@@ -0,0 +1,1333 @@
+        partner_used = timedelta(milliseconds=data[GppProgramProvider._TAKeys.USED_PART_TIME])
+
+        return TimeAllocation(
+            category=category,
+            program_awarded=program_awarded,
+            partner_awarded=partner_awarded,
+            program_used=program_used,
+            partner_used=partner_used)
+
+    def parse_or_group(self, data: dict, program_id: ProgramID, group_id: GroupID) -> OrGroup:
+        """
+        There are no OR groups in the OCS, so this method simply throws a
+        NotImplementedError if it is called.
+        """
+        raise NotImplementedError('OCS does not support OR groups.')
+
+    def parse_and_group(self, data: dict, program_id: ProgramID, group_id: GroupID,
+                        split: bool, split_by_iterator: bool) -> Optional[AndGroup]:
+        """
+        In the OCS, a SchedulingFolder or a program are AND groups.
+        We do not allow nested groups in OCS, so this is relatively easy.
+
+        This method expects the data from a SchedulingFolder or from the program.
+
+        Organizational folders are ignored, so they require some special handling:
+        we retrieve all the observations here that are in organizational folders and
+        simply stick them in this level.
+        """
+        delay_min = timedelta.min
+        delay_max = timedelta.max
+
+        # Get the group name: ROOT_GROUP_ID if the root group and otherwise the name.
+        if GppProgramProvider._GroupKeys.GROUP_NAME in data:
+            group_name = data[GppProgramProvider._GroupKeys.GROUP_NAME]
+        else:
+            group_name = ROOT_GROUP_ID.id
+
+        # Parse notes for "do not split" information if not found previously
+        notes = [(data[key][GppProgramProvider._NoteKeys.TITLE], data[key][GppProgramProvider._NoteKeys.TEXT])
+                 for key in data.keys() if key.startswith(GppProgramProvider._ProgramKeys.NOTE)]
+        if split:
+            split = not self.parse_notes(notes, GppProgramProvider._NO_SPLIT_STRINGS)
+        # Parse notes for "split by interator" information if not found previously
+        if not split_by_iterator:
+            split_by_iterator = self.parse_notes(notes, GppProgramProvider._SPLIT_BY_ITER_STRINGS)
+
+        # Collect all the children of this group.
+        children = []
+
+        # Parse out the scheduling groups recursively.
+        scheduling_group_keys = sorted(key for key in data
+                                       if key.startswith(GppProgramProvider._GroupKeys.SCHEDULING_GROUP))
+        for key in scheduling_group_keys:
+            subgroup_id = GroupID(key.split('-')[-1])
+            subgroup = self.parse_and_group(data[key], program_id, subgroup_id, split=split,
+                                            split_by_iterator=split_by_iterator)
+            if subgroup is not None:
+                children.append(subgroup)
+
+        # Grab the observation data from any organizational folders.
```

Comment:
Do we need to represent org folders in the ODB?

Reply:
Not at the moment.